src/CustomElementInternals.js

```diff
@@ -0,0 +1,294 @@
+   *       - (Element queue pushed to the custom element reactions stack.)
+   *     - Create an element:
+   *       https://dom.spec.whatwg.org/#concept-create-element
+   *       - Sync CE flag is true?
+   *         - Constructor called.
+   *         - Self-modification restrictions enforced.
+   *       - Sync CE flag is false?
+   *         - (Upgrade reaction enqueued.)
+   *     - Attributes appended to element.
+   *       (`attributeChangedCallback` reactions enqueued.)
+   *     - Will execute script flag is true?
+   *       - (Element queue popped from the custom element reactions stack.
+   *         Reactions in the popped stack are invoked.)
+   *   - (Element queue pushed to the custom element reactions stack.)
+   *   - Insert the element:
+   *     https://dom.spec.whatwg.org/#concept-node-insert
+   *     - Shadow-including descendants are connected. During parsing
+   *       construction, there are no shadow-*excluding* descendants.
+   *       However, the constructor may have validly attached a shadow
+   *       tree to itself and added descendants to that shadow tree.
+   *       (`connectedCallback` reactions enqueued.)
+   *   - (Element queue popped from the custom element reactions stack.
+   *     Reactions in the popped stack are invoked.)
+   *
+   * @param {!Node} root
+   * @param {!Set<Node>=} visitedImports
+   */
+  upgradeTree(root, visitedImports) {
+    if (!visitedImports) {
+      visitedImports = new Set();
+    }
+
+    const elements = [];
+
+    const gatherElements = element => {
+      if (element.localName === 'link' && element.getAttribute('rel') === 'import') {
+        if (visitedImports.has(element)) return;
+
+        visitedImports.add(element);
+
+        // The HTML Imports polyfill sets a descendant element of the link to
+        // the `import` property, specifically this is *not* a Document.
+        if (element.import instanceof Node) {
+          Utilities.walkDeepDescendantElements(element.import, gatherElements);
+        } else {
+          element.addEventListener('load', () => {
+            const doc = element.import;
+
+            if (doc[CESymbols.documentLoadHandled]) return;
+            doc[CESymbols.documentLoadHandled] = true;
+
+            this.upgradeTree(document);
+          });
+        }
+      } else {
+        elements.push(element);
+      }
+    };
+
+    Utilities.walkDeepDescendantElements(root, gatherElements);
+
+    for (var i = 0; i < elements.length; i++) {
+      this.upgradeElement(elements[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  upgradeElement(element) {
+    const currentState = element[CESymbols.state];
+    if (currentState !== undefined) return;
+
+    const definition = this.localNameToDefinition(element.localName);
+    if (!definition) return;
+
+    definition.constructionStack.push(element);
+
+    const constructor = definition.constructor;
+    try {
+      try {
+        let result = new (constructor)();
```

Comment:
Do these `try`'s add any perf impact?  Could you just de-FUD by A-B testing with/without?